### ISSUE TITLE
Expand sd_api_pictures to support adventure game/interactive story use-case

### ISSUE THREAD

**Issue by vaidasmaciulis** [oobabooga/text-generation-webui#533]
**Description**

I have modified sd_api_pictures script locally to use it for a graphic text adventure game:

![image](https://user-images.githubusercontent.com/11597256/227478569-b21a9042-9e71-4603-b606-a164cc9259d4.png)

Would be nice to incorporate changes required to support such use case in main repo.

Changes that is required to achieve this (all should be made configurable, to not interfere with current behavior):

- Generate image on every response
- Do not modify input text
- Do not modify output text
- Allow to configure image width/height independently
- Do not resize output image

Modified POC script, that can be used as reference:

```
import base64
import io
import re
from pathlib import Path

import gradio as gr
import modules.chat as chat
import modules.shared as shared
import requests
import torch
from PIL import Image

torch._C._jit_set_profiling_mode(False)

# parameters which can be customized in settings.json of webui  
params = {
    'enable_SD_api': True,
    'address': 'http://127.0.0.1:7860',
    'save_img': False,
    'SD_model': 'NeverEndingDream', # not really used right now
    'prompt_prefix': 'fantasy painting, ',
    'negative_prompt': 'text, caption, words, 3d, nrealfixer, nfixer, nartfixer',
    'side_length': 512,
    'restore_faces': False
}

SD_models = ['NeverEndingDream'] # TODO: get with http://{address}}/sdapi/v1/sd-models and allow user to select

streaming_state = shared.args.no_stream # remember if chat streaming was enabled
picture_response = True # specifies if the next model response should appear as a picture
pic_id = 0

def remove_surrounded_chars(string):
    # this expression matches to 'as few symbols as possible (0 upwards) between any asterisks' OR
    # 'as few symbols as possible (0 upwards) between an asterisk and the end of the string'
    return re.sub('\*[^\*]*?(\*|$)','',string)

# I don't even need input_hijack for this as visible text will be commited to history as the unmodified string
def input_modifier(string):
    """
    This function is applied to your text inputs before
    they are fed into the model.
    """
    # Adventure game: no modification on input
    return string

# Get and save the Stable Diffusion-generated picture
def get_SD_pictures(description):

    global params, pic_id

    payload = {
        "prompt": params['prompt_prefix'] + description,
        "seed": 3463809478,
        "sampler_name": "Euler a",
        "steps": 12,
        "cfg_scale": 7,
        "width": 640, # Adventure game change
        "height": params['side_length'],
        "restore_faces": params['restore_faces'],
        "negative_prompt": params['negative_prompt']
    }
    
    response = requests.post(url=f'{params["address"]}/sdapi/v1/txt2img', json=payload)
    r = response.json()

    visible_result = ""
    for img_str in r['images']:
        image = Image.open(io.BytesIO(base64.b64decode(img_str.split(",",1)[0])))
        if params['save_img']:
            output_file = Path(f'extensions/sd_api_pictures/outputs/{pic_id:06d}.png')
            image.save(output_file.as_posix())
            pic_id += 1
        # Adventure game: no image resize
        buffered = io.BytesIO()
        image.save(buffered, format="JPEG")
        buffered.seek(0)
        image_bytes = buffered.getvalue()
        img_str = "data:image/jpeg;base64," + base64.b64encode(image_bytes).decode()
        visible_result = visible_result + f'<img src="{img_str}" alt="{description}">\n'
    
    return visible_result

# TODO: how do I make the UI history ignore the resulting pictures (I don't want HTML to appear in history)
# and replace it with 'text' for the purposes of logging?
def output_modifier(string):
    """
    This function is applied to the model outputs.
    """
    global pic_id, picture_response, streaming_state

    #if shared.args.no_stream:#not picture_response:
        #return string

    prompt = remove_surrounded_chars(string)
    prompt = string.replace('"', '')
    prompt = string.replace('“', '')
    prompt = string.replace('\n', ' ')
    prompt = string.strip()

    # Adventure game: no modification on output

    image = get_SD_pictures(prompt)

    #picture_response = False

    shared.processing_message = "*Is typing...*"
    #shared.args.no_stream = streaming_state
    return image + "\n" + string

def bot_prefix_modifier(string):
    """
    This function is only applied in chat mode. It modifies
    the prefix text for the Bot and can be used to bias its
    behavior.
    """

    return string

def force_pic():
    global picture_response
    picture_response = True

def ui():

    # Gradio elements
    with gr.Accordion("Stable Diffusion api integration", open=True):
        with gr.Row():
            with gr.Column():
                enable = gr.Checkbox(value=params['enable_SD_api'], label='Activate SD Api integration')
                save_img = gr.Checkbox(value=params['save_img'], label='Keep original received images in the outputs subdir')
            with gr.Column():
                address = gr.Textbox(placeholder=params['address'], value=params['address'], label='Stable Diffusion host address')
        
        with gr.Row():
            force_btn = gr.Button("Force the next response to be a picture")
            generate_now_btn = gr.Button("Generate an image response to the input")

        with gr.Accordion("Generation parameters", open=False):
            prompt_prefix = gr.Textbox(placeholder=params['prompt_prefix'], value=params['prompt_prefix'], label='Prompt Prefix (best used to describe the look of the character)')
            with gr.Row():
                negative_prompt = gr.Textbox(placeholder=params['negative_prompt'], value=params['negative_prompt'], label='Negative Prompt')
                dimensions = gr.Slider(256,702,value=params['side_length'],step=64,label='Image dimensions')
                # model = gr.Dropdown(value=SD_models[0], choices=SD_models, label='Model')
    
    # Event functions to update the parameters in the backend
    enable.change(lambda x: params.update({"enable_SD_api": x}), enable, None)
    save_img.change(lambda x: params.update({"save_img": x}), save_img, None)
    address.change(lambda x: params.update({"address": x}), address, None)
    prompt_prefix.change(lambda x: params.update({"prompt_prefix": x}), prompt_prefix, None)
    negative_prompt.change(lambda x: params.update({"negative_prompt": x}), negative_prompt, None)
    dimensions.change(lambda x: params.update({"side_length": x}), dimensions, None)
    # model.change(lambda x: params.update({"SD_model": x}), model, None)

    force_btn.click(force_pic)
    generate_now_btn.click(force_pic)
    generate_now_btn.click(eval('chat.cai_chatbot_wrapper'), shared.input_params, shared.gradio['display'], show_progress=shared.args.no_stream)
```

**Comment by vaidasmaciulis** [comment]
[Brawlence](https://github.com/Brawlence) Thanks for taking this! Your mentioned 3 modes makes sense to me. I think it can be expanded further as people find new usages for it. 
As for logs I have checked my test game, it has 54 images in original 512x512, and json file is just above 2mb.

**Comment by oobabooga** [comment]
A PR with these updates would be nice. Also, take a look at the thread below, he obtained some nice results:

https://www.reddit.com/r/Oobabooga/comments/120rxna/testing_out_image_recognition_input_techniques/

**Comment by Brawlence** [comment]
@vaidasmaciulis if you want to test the upcoming changes here's the most recent version I've got: [sd-api-pics-improvement](https://github.com/Brawlence/text-generation-webui/tree/sd-api-pics-improvement)

Features: API detection (press enter in the API box), more robust host address handing, VRAM management (model shuffling), separate image dimensions sliders for H&W, three different trigger modes (manual, interactive, always-on). The button to force/suppress the image on the next reply. 
The model input is modified only in the interactive mode; other two are unaffected. The output pic description is presented differently for Picture-book / Adventure mode.

Just make sure to **disable VRAM management** thus far as it requires patches both on `oobabooga` and on `Automatic1111`.

TODO: zoom-in feature for bigger images

@oobabooga I've looked through the changes -- they are mechanically the same as proposed here

**Comment by vaidasmaciulis** [comment]
[Brawlence](https://github.com/Brawlence), I just got everything working after recent updates. Can you create PR to this repo instead of fork? Or you need special access for that?

**Comment by Brawlence** [comment]
Sorry, that was probably the wrong branch. Get the new script.py from #596 , it should be compatible with all the recent changes right away. If you would like to test the VRAM management feature, copy changes to server.py as well (there are small, only several lines)

**Comment by vaidasmaciulis** [comment]
Yes, script from PR seems to work! Wonder if settings can persist between loading UI?

**Comment by vaidasmaciulis** [comment]
[Brawlence](https://github.com/Brawlence), Few things noticed after some testing:

1. There is two titles for plugin
![image](https://user-images.githubusercontent.com/11597256/228041717-c2d49d0e-c8de-43c3-9437-2c68497bfcc3.png)

2. If I generate text in Manual mode, and then switch to Adventure mode, images are not being generated until I click Force generate button.

**Comment by Brawlence** [comment]
@vaidasmaciulis 
> Yes, script from PR seems to work! Wonder if settings can persist between loading UI?

Yup, you can customize the settings to your liking in settings.json. [README.MD describes how to do it](https://github.com/oobabooga/text-generation-webui/tree/main/extensions/sd_api_pictures#persistents-settings)

Also, 
@oobabooga this issue is resolved, I think you can safely close it.
